src/samplers/support/hmc_core.jl

```diff
@@ -11,58 +11,91 @@ end
 
 sample_momentum(vi::VarInfo, spl::Sampler) = begin
   dprintln(2, "sampling momentum...")
-  randn(length(getranges(vi, spl))) ./ spl.info[:wum][:stds]
+  # randn(length(getranges(vi, spl))) ./ spl.info[:wum][:stds]
+
+  d = length(getranges(vi, spl))
+  stds = spl.info[:wum][:stds]
+
+  return _sample_momentum(d, stds)
+
+end
+
+function _sample_momentum(d::Int, stds::Vector)
+
+  return randn(d) ./ stds
+
 end
 
 # Leapfrog step
 # NOTE: leapfrog() doesn't change θ in place!
-leapfrog(_θ::Union{Vector,SubArray}, p::Vector{Float64}, τ::Int, ϵ::Float64,
-          model::Function, vi::VarInfo, spl::Sampler) = begin
+leapfrog(_θ::T, p::Vector{Float64}, τ::Int, ϵ::Float64,
+          model::Function, vi::VarInfo, spl::Sampler) where {T<:Union{Vector,SubArray}} = begin
 
   θ = realpart(_θ)
-  if ADBACKEND == :forward_diff
-    vi[spl] = θ
-    grad = gradient(vi, model, spl)
-  elseif ADBACKEND == :reverse_diff
-    grad = gradient_r(θ, vi, model, spl)
+
+  lp_grad_func(θ::T) where {T<:Union{Vector,SubArray}} = begin
+
+    if ADBACKEND == :forward_diff
```

Comment:
Agree. But it might not be possible to make it into a function because the main reason to introduce this PR and hence this function is to avoid having `vi` in any function signature. But surely I can use a macro to achieve this pattern - will do it for the DRY principle.